### DISCUSSION TITLE
C# syntax like Python `bytes` and `bytearray`

### DISCUSSION THREAD

**Discussion by richlander** [dotnet/csharplang#6475] [Language Ideas]
I was recently reading and porting some Python code to C# for Raspberry Pi/IoT. Most of that was straightforward, but I came across this [`bytes` function](https://docs.python.org/3/c-api/bytes.html) that was new to me. I immediately fell in love with it and saw it as a "DSL for byte arrays" or "stringbuilder for byte arrays". It is incredibly useful for scenarios where you need to hand-craft bytes, like IoT.

This is the primary file I was porting: https://github.com/adafruit/Adafruit_CircuitPython_IS31FL3731/blob/main/adafruit_is31fl3731/__init__.py

Here's a [good example](https://github.com/adafruit/Adafruit_CircuitPython_IS31FL3731/blob/794488d1de3d17d1a08887c4a651cfac2c5a1524/adafruit_is31fl3731/__init__.py#L145):

```python
_ENABLE_OFFSET = const(0x00)
enable_data = bytes([_ENABLE_OFFSET] + [255] * 18)
```

This is saying, give me a byte array with the first value as `0x00` and then 18 bytes as `255`. AFAIK, C# doesn't offer a terse syntax like that.

This [example](https://github.com/adafruit/Adafruit_CircuitPython_IS31FL3731/blob/794488d1de3d17d1a08887c4a651cfac2c5a1524/adafruit_is31fl3731/__init__.py#L114) is also good since it enables adding a byte array as part of a new byte array.

```python
    def _i2c_write_reg(self, reg, data):
        # Write a contiguous block of data (bytearray) starting at the
        # specified I2C register address (register passed as argument).
        self._i2c_write_block(bytes([reg]) + data)

    def _register(self, bank, register, value=None):
        self._bank(bank)
        if value is None:
            result = bytearray(1)
            return self._i2c_read_reg(register, result)[0]
        self._i2c_write_reg(register, bytearray([value]))
        return None
```

This is [how I managed that in C#](https://github.com/richlander/iot/blob/b48722382f3b36ed5f25ad930fdbee67d773a25e/src/devices/Is31Fl3731/Is31Fl3731.cs#L211-L213):

```csharp
        private void Write(byte register, byte address, byte[] value)
        {
            _i2cDevice.Write(new byte[] { COMMAND_REGISTER, register });
            byte[] data = new byte[value.Length + 1];
            data[0] = address;
            value.CopyTo(data.AsSpan(1));
            _i2cDevice.Write(data);
        }
```

From a C# standpoint, this whole space might be thought of as a byte[] serializer driven by a source generator. I don't know. I just like the experience. If there is a better way to accomplish what I'm doing in C#, that'd be great to know.

**Comment by CyrusNajmabadi** [comment]
I really don't like the use of `[...]` to have some sort of special meaning.  You would be able to get close to the above though by defining some custom types and using operator overloading for this purpose.

**Reply by richlander** [reply]
I'm not proposing the specific syntax, but the experience. Perhaps the existing `params` affordance could help.

**Reply by CyrusNajmabadi** [reply]
Could you clarify which parts of the experience you'd like potential improvements for here?  I saw things like teh DSL for `[_ENABLE_OFFSET] + [255] * 18`.  It does seem like you could do something similar in C# (albeit not with the `[...]` syntax), through custom types and operator overloading.  If that's not a good solution, what sort of thing would you be looking for?

**Reply by richlander** [reply]
I like the DSL-like syntax where I can specify the final byte array I want in a way that is intuitive. That said, I don't think that this python is all that intuitive. The meaning of `+` (demarcator) and `*` (repeat) wasn't immediately clear to me. Same with the brackets.

Using the same example (with C#), I would have preferred something closer to:

byte[] data = new byte[]
{
_ENABLE_OFFSET,
255 {repeat:18}
};

With the other example (embedded byte array), one can imagine:

byte[] data = new byte[]
{
address,
value {flatten}
};

I'm far from a syntax expert so don't take any of that as a hard proposal. Main point here is that it would be nice to extend the object initializer syntax. That could probably benefit a lot of scenarios.

**Reply by ufcpp** [reply]
Can we use [collection literals](https://github.com/dotnet/csharplang/issues/5354) for this purpose? e.g.

```cs
var data = [
    _ENABLE_OFFSET,
    ..[ 255 ] * 18 // or "..([ 255 ] * 18)" owing to operator precedence?
];
```

**Reply by CyrusNajmabadi** [reply]
`[ 255 ] * 18` would not give you 18 copies of 255 (nor would i want it to).  You could certainly have some sort of extnesion like `[(byte)255].Repeat(18)`, etc.

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi I think the intention here was that with an operator (or a keyword), the language could optimize it, whereas the `.Repeat` extension would allocate an array, only to then be consumed and not used anymore. Or, if `Repeat` method returns `IEnumerable<T>`, it would also switch to unknown length collection literal, which again precludes optimizations.

**Reply by CyrusNajmabadi** [reply]
I think the easy to do that would be to have a custom type representing the dsl.  Then you only allocate once at the end when you make your bytes.  This would be possible today without needing lang support.

**Comment by HaloFour** [comment]
IIRC `bytes` and `bytearray` are just normal methods in Python.  It sounds like you're more interested in the syntax for defining and manipulating lists, where `[x]` defines a list literal, `+` concats lists and `*` repeats a list.

**Reply by tfenise** [reply]
It's not obvious to me why `+` should be concating a list, and why `*` should be repeating a list. Why couldn't `[1, 2] * 3` be interpreted as `[1 * 3, 2 * 3]`, rather than `[1, 2, 1, 2, 1, 2]`? There are probably a lot of libraries of linear algebra doing similar things.

**Comment by En3Tho** [comment]
F# has collection builders (example with array)

```fsharp
let bytes = [| const; yield! Enumerable.Repeat(255uy, 18) |] // or F#'s Array.init or whatever
// or
let bytes = [| const; for _ = 1 to 18 do 255uy |]
```
Or you can make a.custom computation expression so similar kind of syntax can work with List, HashSet and other collection types 

I believe the closest thing in C# are collection initializers. But they can't really yield a collection can they?

Maybe something like J's/TS can work?
```csharp
byte[] x = new byte[] { const, ..bytes }
```

As for the original question I guess extension methods can give a somewhat nice feel, but they might be not be effective.

**Comment by tfenise** [comment]
> This is saying, give me a byte array with the first value as 0x00 and then 18 bytes as 255. AFAIK, C# doesn't offer a terse syntax like that.

`new byte[1] { 0x00 }.Concat(Enumerable.Repeat<byte>(0xFF, 18)).ToArray()`

**Reply by FaustVX** [reply]
Or even better (at least to me, as it doesn't allocate an array) : `Enumerable.Repeat<byte>(0xFF, 18).Prepend((byte)0x00).ToArray();`

**Reply by Joe4evr** [reply]
But what seems to be asked for here is that the compiler could generate the full array initializer itself so there's only a single allocation at runtime, at least for the case of constant inputs. Specifically, if the case of constant inputs can be enforced in some way, you immediately have compiler verification that all the inputs *are* in fact constants (like what #5295 talks about).

**Comment by TahirAhmadov** [comment]
I think this can be worked into collection literals https://github.com/dotnet/csharplang/issues/5354:
```cs
byte[] buf = [0x00, ..((byte)0xFF).Repeat(18)];
byte[] buf = [0x00, ..0xFFt.Repeat(18)]; // here, I invented a suffix, "t", which means "byTe"
byte[] buf = [0x00, ..([0xFFt] * 18)];
```

**Comment by HaloFour** [comment]
Theoretically with list literals and extension everything you could get that same functionality without other language features.  It wouldn't likely be terribly efficient, though.  I kind of doubt that it's efficient in Python as you're concatenating lists of integers and then calling a method to convert that to a byte array, but maybe Python recognizes the pattern and optimizes it.

**Reply by ufcpp** [reply]
Some target-typed syntax can not use extension methods, for instance:

```cs
IFormattable f = $"";
f.M(); // OK

$"".M(); // error

static class Ex
{
    public static void M(this IFormattable f) { }
}
```

[List literals](https://github.com/dotnet/csharplang/issues/5354) might be target-typed? So, can they use [extension everything](https://github.com/dotnet/csharplang/issues/192)?

```cs
new[] { 255 }.Repeat(18); // OK

[255].Repeat(18); // should be OK?

static class Ex
{
    public static IEnumerable<T> Repeat<T>(this IEnumerable<T> values, int count)
    {
        for (int i = 0; i < count; i++)
        {
            foreach (var value in values)
            {
                yield return value;
            }
        }
    }
}
```

**Reply by CyrusNajmabadi** [reply]
> [255].Repeat(18); // should be OK?

you'd need:

`[(byte)255].Repeat(18);`

Or you'd need the receiver of hte extension to be something like `this IEnumerable<byte> values`.

**Comment by Joe4evr** [comment]
I mentioned https://github.com/dotnet/csharplang/issues/5295 here earlier and I do believe that it would be beneficial if that and this can be considered together. By which I mean: if a little bit of syntax needs to be invented to make this feature appealing (which by default could support building an array from *arbitrary* values at runtime, even if that requires multiple allocations, but may still be more optimizable than going through LINQ), there should be a consistent transformation the developer can apply to make the compiler *enforce* that all inputs are compile-time constant (and therefor optimizable to a single array initializer at runtime).

Parity between those 2 scenarios would have *enormous* value IMO, otherwise you end up with one of them requiring the removal of some keyword where the other requires *adding* some keyword in order to get the same "(supported) constant input only" behavior (which is  only going to be another little wart). Even if this feature isn't considered for inclusion *now*, it's something that still can be brought to influence the ultimate design of 5295.
